lib/openstax/book.rb

```diff
@@ -0,0 +1,83 @@
+require_relative './page'
+
+module Openstax
+  # A Openstax::Book is the object that "represents" the book to the search
+  # world.  A book is composed of pages, and pages are composed of elements.
+  #
+  # An indexing stategy is injected in order to determine what elements we
+  # parse out of the page.
+  #
+  # Pages and Elements are lazy loaded.
+  class Book
+    CNX_URL_PREFIX = 'https://archive.cnx.org/contents'
+
+    attr_reader :uuid, :version
+
+    def initialize(uuid:, version:, indexing_strategy:)
+      @uuid = uuid
+      @version = version
+      @indexing_strategy = indexing_strategy
+    end
+
+    def pages
+      @pages ||= begin
+        page_ids = recursive_flatten_page_ids(data['tree']['contents'])
+        Rails.logger.info "OpenSearch: found #{page_ids.count} possible page_ids in book #{@uuid}@#{@version}"
+
+        pages = page_ids.map do |page_id|
+          response = fetch_cnx_url(page_uri(page_id)) rescue next
+
+          Page.new(id: page_id,
+                   data: JSON.parse(response.body),
+                   desired_page_elements: @indexing_strategy.desired_page_elements)
+        end.compact
+
+        Rails.logger.info "OpenSearch: Loaded book #{@uuid}@#{@version} of #{pages.count} pages"
+        pages
+      end
+    end
```

Comment:
If some of the other code reorg happens and the traversal of books happens more in the indexing strategy, it may make sense to add methods to these book/chapter/page/etc classes that use a visitor pattern, accepting the index strategy to use or ignore parts of the book -- that's a little what's starting to happen already except the strategy is passed during initialization instead of during a `visit` method that crawls an existing tree.